packages/remix-dev/config.ts

```diff
@@ -375,41 +369,66 @@ export async function readConfig(
     }
   }
 
-  let isCloudflareRuntime = ["cloudflare-pages", "cloudflare-workers"].includes(
-    appConfig.serverBuildTarget ?? ""
-  );
-  let isDenoRuntime = appConfig.serverBuildTarget === "deno";
+  // TODO: default to esm for v2
+  let serverModuleFormat: ServerModuleFormat = "cjs";
+  let serverPlatform: ServerPlatform = "node";
+  let serverConditions = undefined;
+  let serverMainFields = undefined;
+  let serverBuildPath = "build/index.js";
+  let defaultPublicPath = "/build/";
+  let serverDependenciesToBundle: "all" | Array<string | RegExp> | undefined;
+
+  // Retain backwards compatibility for `serverBuildTarget`
+  switch (appConfig.serverBuildTarget) {
+    case "arc":
+      serverBuildPath = "server/index.js";
+      defaultPublicPath = "/_static/build/";
+      break;
+    case "cloudflare-pages":
+      serverBuildPath = "functions/[[path]].js";
+    case "cloudflare-workers":
+      serverConditions = ["worker"];
+      serverModuleFormat = "esm";
+      serverPlatform = "neutral";
+      serverConditions = ["worker"];
+      serverMainFields = ["browser", "module", "main"];
+      serverDependenciesToBundle = "all";
+      break;
+    case "deno":
+      serverModuleFormat = "esm";
+      serverPlatform = "neutral";
+      serverConditions = ["deno", "module", "main"];
+      serverMainFields = ["module", "main"];
+      serverDependenciesToBundle = "all";
+      break;
+    case "netlify":
+      serverBuildPath = ".netlify/functions-internal/server.js";
+      break;
+    case "vercel":
+      serverBuildPath = "api/index.js";
+      break;
+  }
+
+  // override with new config if provided
+  serverConditions = appConfig.serverConditions
+    ? appConfig.serverConditions
+    : serverConditions;
+  serverMainFields = appConfig.serverMainFields
+    ? appConfig.serverMainFields
+    : serverMainFields;
+  serverDependenciesToBundle =
+    (appConfig.serverDependenciesToBundle
+      ? appConfig.serverDependenciesToBundle
+      : serverDependenciesToBundle) || [];
```

Comment:
This can be simplified to

```suggestion
  serverConditions = appConfig.serverConditions ?? serverConditions;
  serverMainFields = appConfig.serverMainFields ?? serverMainFields;
  serverDependenciesToBundle =
    appConfig.serverDependenciesToBundle ?? serverDependenciesToBundle;
```

Reply:
We also need to override `serverModuleFormat` & `serverPlatform` with config + **always** set `serverMainFields` to `["module", "main"]` if `serverModuleFormat` is set to `"esm"`